### PR TITLE
rtaudio_6: init at 6.0.1

### LINKED ISSUE #274031: Update request: rtaudio 5.2.0 → 6.0.1
- Package name:
- Latest released version: 5.2.0
- Current version on the unstable channel: 5.2.0
- Current version on the stable/release channel: 5.2.0

**Notify maintainers**

@magnetophon 

-----

Note for maintainers: Please tag this issue in your PR.

### PR DESCRIPTION
**Pull request by OPNA2608**
## Description of changes

Closes #274031

Due to RtAudio 6 being an API-breaking update, some reverse dependencies' upstreams being quite slow to adapt to it, and others' updates being blocked by the `rtaudio` version getting held back, introduce a separate `rtaudio_6` for reverse dependencies to opt into. As an example, also switch `bambootracker{,-qt6}` from its vendored RtAudio 6 to the new `rtaudio_6` to check that it works.

I'm not sure if making v6 the default should be included in 24.05, so I haven't made any treewide adjustments here.

Pinging some involved ppl from #245075: @PowerUser64 @lilyinstarlight @luzpaz @eclairevoyant 

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by Astavie — APPROVED** [review]

File: pkgs/by-name/rt/rtaudio_6/package.nix [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+  nativeBuildInputs = [ cmake pkg-config ];
+
+  buildInputs = lib.optional alsaSupport alsa-lib
+    ++ lib.optional pulseaudioSupport libpulseaudio
+    ++ lib.optional jackSupport jack
+    ++ lib.optional coreaudioSupport CoreAudio;
+
+  cmakeFlags = [
+    "-DRTAUDIO_API_ALSA=${if alsaSupport then "ON" else "OFF"}"
```
Comment (eclairevoyant) [inline pkgs/by-name/rt/rtaudio_6/package.nix:36]: could be simplified a lot / be less error-prone with the cmake functions in `lib`

```suggestion
    (lib.cmakeBool "RTAUDIO_API_ALSA" alsaSupport)
```

same with the others

File: pkgs/by-name/rt/rtaudio_6/package.nix [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+}:
+
+stdenv.mkDerivation rec {
+  pname = "rtaudio";
+  version = "6.0.1";
+
+  src = fetchFromGitHub {
+    owner = "thestk";
+    repo = "rtaudio";
+    rev = version;
```
Comment (eclairevoyant) [inline pkgs/by-name/rt/rtaudio_6/package.nix:24]:
```suggestion
    rev = "refs/tags/${version}";
```

prevents confusing the update bots/scripts

File: pkgs/by-name/rt/rtaudio_6/package.nix [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    sha256 = "11llxdg62gs6l2zc473s9zfb8jczflyq1dd0k0wfcmvyg5p33jq1";
```
Comment (eclairevoyant) [inline pkgs/by-name/rt/rtaudio_6/package.nix:25]: SRI hash would be preferred

File: pkgs/by-name/rt/rtaudio_6/package.nix [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    ++ lib.optional coreaudioSupport CoreAudio;
+
+  cmakeFlags = [
+    "-DRTAUDIO_API_ALSA=${if alsaSupport then "ON" else "OFF"}"
+    "-DRTAUDIO_API_PULSE=${if pulseaudioSupport then "ON" else "OFF"}"
+    "-DRTAUDIO_API_JACK=${if jackSupport then "ON" else "OFF"}"
+    "-DRTAUDIO_API_CORE=${if coreaudioSupport then "ON" else "OFF"}"
+  ];
+
+  meta = with lib; {
```
Comment (eclairevoyant) [inline pkgs/by-name/rt/rtaudio_6/package.nix:42]:
```suggestion
  meta = {
```

tracking issue: #292468

**Comment by eclairevoyant** [comment]
also looks like bambootracker fails on darwin due to some warnings-as-errors:

```
In file included from instrument/instruments_manager.cpp:35:
./utils.hpp:96:26: error: 'result_of<(lambda at instrument/instruments_manager.cpp:733:2) &&(std::shared_ptr<AbstractInstrument>)>' is deprecated [-Werror,-Wdeprecated-declarations
```

**Comment by OPNA2608** [comment]
Re: suggestions: I agree with all of them, I just started by copy-pasting our v5 expression to keep the differences to the currently-maintained to a minimum. I guess this would be a good-enough occasion it modernise it abit...

> also looks like bambootracker fails on darwin due to some warnings-as-errors:

*Arrrgh*, I'll try to fix this upstream when I have time. I remember Fedora opened a similar ticket recently...

**Comment by OPNA2608** [comment]
> also looks like bambootracker fails on darwin

This is `bambootracker-qt6` specific, and already failing on Hydra: https://hydra.nixos.org/eval/1805974?filter=bambootracker&compare=1805969. Do you think fixing that here would be relevant enough, to show that `rtaudio_6` works everywhere? Otherwise I'd handle this along with any other failures of my packages during ZHF.

**Comment by Astavie** [comment]
make sure to squash all your commits into a single commit in the end

**Comment by OPNA2608** [comment]
I could see squashing the modernisation into the init, if that's what you mean. Meddling with other packages should prolly stay in a separate commit though.

**Comment by Astavie** [comment]
I seem to remember someone telling me all commits should be squashed, but looking at contributing.md it appears I was wrong, apologies!

**Review by eclairevoyant — APPROVED** [review]
One commit = one logical change. I'd say the commits are fine as-is.

> This is `bambootracker-qt6` specific, and already failing on Hydra: https://hydra.nixos.org/eval/1805974?filter=bambootracker&compare=1805969. Do you think fixing that here would be relevant enough, to show that `rtaudio_6` works everywhere? Otherwise I'd handle this along with any other failures of my packages during ZHF.

I see, if it's already failing then there's no net downsides to merging as-is. Didn't notice the multiple variants, bambootracker itself appears to build fine; nor did I know about the `validatePkgConfig` hook, that's more reassuring with that added :+1:

**Review by magnetophon — APPROVED** [review]
lgtm.

BambooTracker runs here.

Result of `nixpkgs-review pr 306946` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>bambootracker</li>
    <li>bambootracker-qt6</li>
    <li>rtaudio_6</li>
  </ul>
</details>

**Pull request merged**